src/identity/customca.rs

```diff
@@ -0,0 +1,129 @@
+use super::CertificateProvider;
+use async_trait::async_trait;
+use kube::{
+    api::{Api, PostParams, ListParams, WatchEvent},
+    client::Client,
+};
+use tracing::{instrument, warn};
+use crate::{identity::manager::Identity, tls::SanChecker};
+use crate::identity::Error;
+use crate::tls;
+use futures::{StreamExt, TryStreamExt};
+use k8s_openapi::{api::certificates::v1::{CertificateSigningRequest, CertificateSigningRequestSpec}, ByteString};
+use std::str;
+use rand::prelude::*;
+
+#[derive(Clone)]
+pub struct CustomSigner {
+    pub signer: Option<std::string::String>,
+}
+
+impl CustomSigner {
+    pub fn new(signer: Option<std::string::String>) -> Result<CustomSigner, Error> {
+        Ok(CustomSigner { signer })
+    }
+}
+
+#[async_trait]
+impl CertificateProvider for CustomSigner {
+    #[instrument(skip_all)]
+    async fn fetch_certificate(&self, id: &Identity) -> Result<tls::Certs, Error> {
+        let client: Client = Client::try_default()
+        .await
+        .expect("Expected a valid KUBECONFIG environment variable.");
+        let csr: Api<CertificateSigningRequest> = Api::all(client.clone());
+        let pp = PostParams::default();
+        let cs = tls::CsrOptions {
+            san: id.to_string(),
+        }
+        .generate()?;
+        let cert: Vec<u8> = cs.csr;
+        let pkey = cs.pkey;
+        let csr_name = generate_csr_name();
+        let stored_csr_name = csr_name.clone();
+        let signer_name = &self.signer;
+        let mut usage = Vec::new();
+        usage.push("client auth".to_string());
+
+        let obj_metadata = k8s_openapi::apimachinery::pkg::apis::meta::v1::ObjectMeta{
+            name: Some(csr_name),
+            ..Default::default()
+        };
+        let csr_spec = CertificateSigningRequestSpec{
+            expiration_seconds:Some(86400),
+            request:k8s_openapi::ByteString(cert),
+            signer_name:signer_name.as_ref().unwrap().to_string(),
+            usages:Some(usage),
+            ..Default::default()
+        };
+
+        let csr_req = CertificateSigningRequest {
+            metadata: obj_metadata,
+            spec: csr_spec,
+            ..Default::default()
+        };
+
+        // creating CSR
```

Comment:
Who is going to approve this csr?

Reply:
users need to set their custom signer - https://github.com/istio/ztunnel/blob/4e5667f8ed0ef7ccd3cdfe8b6c67774fc0a62fbd/src/config.rs#L93